### PR TITLE
Add command line tool for checking and repairing corruption

### PR DESCRIPTION
**Pull request by suzaku**
### Check List 

Tests 

 - Manual test

Code changes

Side effects

Related changes

### CODE REVIEW

File: cmd/vlog_health_checker/main.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	}
+	log.Info("Parse fid", zap.Uint32("fid", fid))
+	logFile, err := storage.NewLogFile(fid, path)
+	if err != nil {
+		log.Fatal("Failed to open log", zap.Error(err), zap.String("path", path))
+	}
+	log.Info("Log opened")
+
+	// Open LevelDB
+	metadb, err := storage.OpenMetadataDB(*kvDir, nil)
```
Comment (july2993) [inline cmd/vlog_health_checker/main.go]: close db before quit

File: pump/storage/vlog.go [hunk 2]
```diff
@@ -480,3 +478,13 @@ func (vlog *valueLog) gcTS(gcTS int64) {
 		logFile.lock.Unlock()
 	}
 }
+
+// ParseFid parses the given file name for the Fid.
+func ParseFid(fName string) (uint32, error) {
+	fName = strings.TrimSuffix(filepath.Base(fName), fileExt)
+	fid64, err := strconv.ParseUint(fName, 10, 32)
```
Comment (GregoryIan) [inline pump/storage/vlog.go]: why `fid64`, not `fid32`？

File: cmd/vlog_health_checker/main.go [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+	defer func() {
+		err := metadb.Close()
```
Comment (kennytm) [inline cmd/vlog_health_checker/main.go]: Put the defer *after* `err != nil`...

File: cmd/vlog_health_checker/main.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+				return nil
+			}
+
+			pointer, err := getSavedPointer(metadb, key)
+			if err != nil {
+				log.Error("Failed to get saved data", zap.Error(err), zap.Binary("key", key))
+				return nil
+			}
+
+			if pointer.Fid != vp.Fid || pointer.Offset != vp.Offset {
```
Comment (GregoryIan) [inline cmd/vlog_health_checker/main.go]: pump client would write one same binlog more than once, if offsset is mismatch, maybe another same binlog is matched

File: pump/storage/vlog.go (hunk 2 above)
Comment (suzaku) [reply pump/storage/vlog.go]: Copied from the original implementation.

**Comment by suzaku** [comment]
@july2993 PTAL

File: cmd/vlog_health_checker/main.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	var corruptionPos int64 = -1
+	checked := make(map[int64]struct{}, 64)
+	logFile.Scan(
+		0,
+		func(vp storage.ValuePointer, record *storage.Record) error {
+			bl, err := record.GetBinlog()
+			if err != nil {
+				log.Fatal("Failed to get binlog", zap.Int64("offset", vp.Offset), zap.Error(err))
+			}
+			// Ignore rollback binlogs
```
Comment (july2993) [inline cmd/vlog_health_checker/main.go]: ignore those only commit ts = 0? may some fake binlog using rollback and commit ts = start ts

File: cmd/vlog_health_checker/main.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+				log.Error("Failed to get saved data", zap.Error(err), zap.Binary("key", key))
+				return nil
+			}
+
+			if pointer.Fid != fid {
+				// Reading a different Fid, this should never happen
+				log.Fatal("Fid mismatch", zap.Uint32("get", pointer.Fid), zap.Uint32("expected", fid))
+			}
+
+			if pointer.Offset != vp.Offset {
```
Comment (july2993) [inline cmd/vlog_health_checker/main.go]: s/if/else if/

Comment (suzaku) [reply cmd/vlog_health_checker/main.go]: The `if` block above called `log.Fatal` which called `os.Exit(1)` for us.

File: cmd/vlog_health_checker/main.go (hunk 5 above)
Comment (suzaku) [reply cmd/vlog_health_checker/main.go]: OK, I'll fix it.

File: cmd/vlog_health_checker/main.go [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+			}
+			tso := getTSO(bl)
+			if _, ok := checked[tso]; ok {
+				log.Info("Skip duplicate binlog", zap.Int64("tso", tso))
+				return nil
+			}
+			checked[tso] = struct{}{}
+			key := encodeKey(bl)
+
+			if *repair {
```
Comment (july2993) [inline cmd/vlog_health_checker/main.go]: better put this after L110, only write it if mismatch

**Comment by GregoryIan** [comment]
I will close it firstly

**Pull request closed without merging**